Tool guidance (#2384)
Azure/azure-sdk-tools: tools/CONTRIBUTING.md

> # Azure SDK Tools Contribution Guidelines

Reviewer: It is worth calling out that we allow pretty much any language for the tools.

Author: Added in Line #477

> This is the repository of tool libraries used by Azure SDK purpose.

Reviewer: ```suggestion
This is the repository used to host tools and libraries used by Azure SDK team engineering system. 
```

Reviewer 2: This is the repository used to host tools and libraries used by _the_ Azure SDK team engineering system.

> For every newly developing tool, create a new folder with name of brief description of the tool. 

Reviewer 2: I would suggest putting this under a new markdown header/section like **Adding a new tool**.

> Every tool should serve for single purpose.

Reviewer: ```suggestion
```

Do not add third party tools directly in your folder, use it as dependencies.

> 

Reviewer: tools should have entries inside the codeowners file.

Author: Add code owner section

## README

> Please add README file for every tool to illustrate:

Reviewer: ```suggestion
Please add README.md file for every tool to illustrate:
```

* The purpose of the tool.

> * Prerequsites before use

Reviewer 2: ```suggestion
* Prerequisites before use
```

* How to use, test and maintain the tool.

Example: [README](https://github.com/Azure/azure-sdk-tools/blob/main/tools/http-fault-injector/README.md) 

## Testing

Please provide certain test cases to cover important workflow, especially the tool using in azure pipelines or running in prod.

Example: [Test library](https://github.com/Azure/azure-sdk-tools/tree/main/tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness.Tests) 

If there is any bundle script, please provide end-to-end test on script as well.

Example: [Custom Test on ci.yml](https://github.com/Azure/azure-sdk-tools/blob/main/tools/code-owners-parser/ci.yml#L35)

## Release ci.yml

For the tool which is publishing to public repository, please provide ci.yml for building, testing and releasing. 

> Example: [ci.yml](https://github.com/Azure/azure-sdk-tools/blob/main/tools/CreateRuleFabricBot/ci.yml)

Reviewer: pipelines should use common templates from eng/pipelines/templates.

Please add comments about the naming convention of the pipelines `tools - <tool-name> - ci` for the public builds and `tools - <tool-name>` for internal builds. 

We should also include notes about only including release steps for internal builds, and conditioning those steps similar to https://github.com/Azure/azure-sdk-tools/blob/main/eng/pipelines/templates/stages/archetype-sdk-tool-dotnet.yml#L89.

Author: Called out below.

> # Azure SDK Tools Contribution Guidelines

Reviewer: We should put this in the root of the repo.

```
/tools/<tool-name>/ @owner1 @owner2
```

## README

Please add README.md file for every tool to illustrate:
* The purpose of the tool.
* Prerequisites before use
* How to use, test and maintain the tool locally and remotely. 
* Better to include where the tool is being used.

Example: [README](https://github.com/Azure/azure-sdk-tools/blob/main/tools/http-fault-injector/README.md) 

## Testing

Please provide certain test cases to cover important workflow, especially the tool using in azure pipelines or running in prod.

Example: [Test library](https://github.com/Azure/azure-sdk-tools/tree/main/tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness.Tests) 

If there is any bundle script, please provide end-to-end test on script as well.

Example: [Custom Test on ci.yml](https://github.com/Azure/azure-sdk-tools/blob/main/tools/code-owners-parser/ci.yml#L35)

## Release ci.yml

- For the tool which is publishing to public repository, please provide ci.yml for building, testing and releasing. 

Example: [ci.yml](https://github.com/Azure/azure-sdk-tools/blob/main/tools/CreateRuleFabricBot/ci.yml)

- To publish the tools, use Azure SDK DevOp feeds:

> Example: [Feeds](https://dev.azure.com/azure-sdk/public/_packaging?_a=feed)

Reviewer: This isn't the correct feed and I think we generally should publish to the feed that is most appropriate for the tool we are creating. For example most of the .NET based tools are published to our azure-sdk-for-net feed, but npm based tools will need to go to npm itself. We should probably define those default feeds in our shared publishing templates as we create them.
